Simple Javascript Clock (Workshop Bounty) (#1300)
hackclub/hackclub: workshops/simple_clock/README.md

---
name: 'JavaScript Clock'

> description: 'Build a clock using HTML, CSS, & JavaScript.'

Reviewer: ```suggestion
description: 'Build a clock using HTML, CSS, & JavaScript'
```

[![homepage](https://cloud-azpvjrywa.vercel.app/image.png)](https://js-css-clock.faisalsayed1.repl.co/)

Here's the [live demo][final_live_demo] and [final code][final_code].

[final_live_demo]: https://js-css-clock.faisalsayed1.repl.co/
[final_code]: https://repl.it/@FaisalSayed1/JS-CSS-Clock
[repl_it]: https://repl.it

## Part 1: Prerequisites
You should have a beginner understanding of:
- HTML
- CSS
- JavaScript

## Part 2: Setup

### Setting up your code environment on Repl.it

[Repl.it](https://repl.it) is an online code editor where you can build your clock. You don't have to use Repl.it but I suggest you do as it sets everything up for you and you don't require any installations.

To get started, go to [https://repl.it/languages/html](https://repl.it/languages/html). Your coding environment will spin up in just a few seconds!

You should see something like the following:

![Image](https://cloud-p7qnbqzo6.vercel.app/image.png)

## Part 3: Inspecting The Default Files

![Image](https://media.giphy.com/media/3o7bufkPz3LRof205G/giphy.gif)

### 1) The HTML file

Let us first take a look at the HTML file (`index.html`).

> At line 1 we have `<!DOCTYPE html>` This declares that this file is a HTML file. If we take a look in the `<html>` tag, we will find a `<body>` tag. Here's where we will write the code. Make sure that all the code you write is included between the opening and closing body tags. (`<body></body>`).

Reviewer: ```suggestion
At line 1 we have `<!DOCTYPE html>` This declares that this file is an HTML file. If we take a look in the `<html>` tag, we will find a `<body>` tag. Here's where we will write the code. Make sure that all the code you write is included between the opening and closing body tags. (`<body></body>`).
```

Make sure to remove line 5 (`<meta name="viewport" content="width=device-width">`) which may cause some weird responsiveness issues for our workshop.

![Image](https://cloud-8h5wlpd5b.vercel.app/metagif.gif)

### 2) The CSS file

This is the `style.css` file in your directory. Whatever styles we give to our website is written in this file. If you take a look in the `<head>` tag in your HTML, you will find a line of code (`<link href="style.css" rel="stylesheet" type="text/css" />`). This means that your HTML file is properly linked to your CSS file!

### 3) The JavaScript file

This is the `script.js` file in your directory. It's empty for now, but we will fill it eventually through the workshop! The code written here will be responsible for the functioning of our clock.

If you take a look in the `<body>` tag, you'll find:

```html
<script src="script.js"></script>
```
This means that your HTML file is linked to your JavaScript.

## Part 4: Building the Clock

### 1) HTML

> Let us start building our clock! First we'll write the HTML code required.

Reviewer: ```suggestion
Let us start building our clock! First, we'll write the HTML code required.
```

Inside of our `<body>` tag, add a `<div>` tag with a class `clock`. Add another `<div>` tag inside of your existing `<div>` and give it a class of `clock-face`.

Now add 3 more `<div>` tags nested inside of the `clock-face` div. Give a class of `hand hour-hand`, `hand min-hand`, `hand second-hand` to each of the 3 new `<div>` tags respectively.

> **Note:** Make sure the `<script>` tag, which is in your body, is below the `<div>` tags or else it will throw you errors.

Reviewer: ```suggestion
**Note:** Make sure the `<script>` tag, which is in your body, is below the `<div>` tags, or else it will throw you errors.
```

  </head>
  <body>
    <div class="clock">
      <div class="clock-face">
        <div class="hand hour-hand"></div>
        <div class="hand min-hand"></div>
        <div class="hand second-hand"></div>
      </div>
    </div>
    <script src="script.js"></script>
  </body>
</html>
```

And that's it! We have completed writing our HTML.

### 2) CSS

After we have done writing the HTML, you'll notice that when we click `RUN`, you just see a white screen. Let's now add some styles to our elements!

Go to `style.css` in your repl.it directory. Let's start by adding some styles to our body! uhh... I meant `<body>`! Add the following code to your CSS:

```css
body {
  background: #01ed7f;
  margin: 0;
  padding: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 100vh;
}
```

**Explanation:**
3 main steps in writing CSS:
1. Select the element / class / id on which you want to apply your styles.
2. Open curly brackets `{}`.
3. Start adding your styles inside the curly brackets!

We start by giving our website a soothing green color. Go ahead and try changing it to the color of your own choice!

> Then we removed the default margin and padding from all the elements in our website (some elements have a default value for their margin and padding such as 2.4, 3 etc). We changed the default display `display: block` to `display: flex`.

Reviewer: ```suggestion
Then we removed the default margin and padding from all the elements in our website (some elements have a default value for their margin and padding such as 2.4, 3, etc). We changed the default display `display: block` to `display: flex`.
```

A very good reason for this is because `flex` enables you to align your items perfectly to the center! And that's what we have done in the next two lines. The `align-items` property aligns all your children elements to center, but horizontally. The `justify-content` property aligns all your children elements to center, but vertically.

> Next we have the `min-height` property which sets a minimum height of `100vh` (viewport height) to your `<body>`. The `min-height` property works together with the `align-items` property and helps the children to align properly at the center.

Reviewer: ```suggestion
Next, we have the `min-height` property which sets a minimum height of `100vh` (viewport height) to your `<body>`. The `min-height` property works together with the `align-items` property and helps the children to align properly at the center.
```

Learn more about [viewport heights](https://www.sitepoint.com/css-viewport-units-quick-start/).

Next, we'll need to add some styles to our clock! Following the 3 main steps given above, let's accomplish this task!

1. Select the class designated to the clock div.

![Image](https://cloud-namz2to6v.vercel.app/image.png)

2. Open curly brackets to write the stylings.

![Image](https://cloud-b7yg1ciea.vercel.app/image.png)

3. Add the following styles to this class.

```css
.clock {
  width: 25rem;
  height: 25rem;
  border: 20px solid whitesmoke;
  border-radius: 50px;
  background: blanchedalmond;
  background-image: url(https://bit.ly/2RcERUw);
  background-size: cover;
  box-shadow: 0 0 0 4px rgba(0, 0, 0, 0.1),
              inset 0 0 0 3px #efefef,
              inset 0 0 10px black,
              0 0 10px rgba(0, 0, 0, 0.2);
}
```

**Explanation**: 

> We gave a width and a height to our clock. Next, we gave it a border and the radius for the border. Next, is the background color for that clock (if you don't like it, consider changing it or even removing it).

Reviewer: ```suggestion
We gave the width and height to our clock. Next, we gave it a border and the radius for the border. Next, is the background color for that clock (if you don't like it, consider changing it or even removing it).
```

Similarly, there's this line:

```css
background-image: url(https://bit.ly/2RcERUw);
```
That image is the face of the clock. we need to add `background-size: cover;` property, in order to show our image properly on the website screen.

Also, we have added some shadows to the element. We gave multiple shadows so as to get a cool effect on our clock. Also, the `inset` property changes the shadow from an outer shadow (`outset`) to an inner shadow.

> Now let's test our code! Click on the **RUN** button and Voila! Here's the basic background for our clock! (if it doesn't looks like this, make sure you crosscheck every single line of code).

Reviewer: ```suggestion
Now let's test our code! Click on the **RUN** button and Voila! Here's the basic background for our clock! (if it doesn't look like this, make sure you crosscheck every single line of code).
```

  transition-timing-function: cubic-bezier(0.1, 2.7, 0.58, 1);
}

.hour-hand {
  width: 30%;
  right: 50%;
}

.min-hand {
  width: 35%;
  right: 50%;
}

.second-hand {
  width: 40%;
  right: 50%;
}
```

**Explanation:** We give the clock hands a black color. We specify the clock hands' height and position.

Next, we give it a property of `top` to `50%`, basically it aligns the clock hands in the very center of our clock. To ensure that our clock rotates properly, we give the clock hands a `transform-origin: 100%;`. The transformation origin is the point around which a transformation is applied.

![Image](https://cloud-e99m314xu.vercel.app/image.png)

Next we added more transition properties such as:

```css
transition: all 0.05s;
transition-timing-function: cubic-bezier(0.1, 2.7, 0.58, 1);
```

The `transition: all 0.05s;` property sets a 0.05s animation duration. The `transition-timing-function: cubic-bezier(0.1, 2.7, 0.58, 1);` property is just a transition effect applied to the clock!

Learn more about [transform-origin](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-origin).
Learn more about [cubic-bezier](https://www.w3schools.com/cssref/func_cubic-bezier.asp).

> **Note:** The `.hand` class was given to 3 `<div>` tags which means this styling will affect all the 3 `<div>` tags. To give some extra classes to the hour hand, minute hand and the second hand specifically, we also gave those 3 `<div>` tags a unique classname.

Reviewer: ```suggestion
**Note:** The `.hand` class was given to 3 `<div>` tags which means this styling will affect all the 3 `<div>` tags. To give some extra classes to the hour hand, minute hand, and the second hand specifically, we also gave those 3 `<div>` tags a unique class name.
```

To align each hand properly, we give them a property of `right: 50%;`.

Here's what we have so far:

![Image](https://cloud-fprpa1onx.vercel.app/image.png)

**What we did so far:**  
- We built the structure of the clock with HTML.  
- We built the display of the clock with CSS.  
- Now with JavaScript, we will build the functionality of the clock.

![Image](https://media.giphy.com/media/oymRJRRiiPaVzDnIF1/giphy.gif)

## 3) JavaScript

> So far, we have the clock hands aligned correctly and now in this section we will animate them. Before panicking, let me ensure you that there's just 30 lines of JavaScript code!

Reviewer: ```suggestion
So far, we have the clock hands aligned correctly and now in this section, we will animate them. Before panicking, let me ensure you that there are just 30 lines of JavaScript code!
```

Go to your `script.js` file. First let's link the `<div>` tags in our HTML to JavaScript. Type the following code:

```js
const secondHand = document.querySelector(".second-hand");
const minHand = document.querySelector(".min-hand");
const hourHand = document.querySelector(".hour-hand");
```

**Explanation:**
We declare the variables `secondHand`, `minHand`, `hourHand` using `const` and set their value to the first element that matches a specified CSS class(s) in the document. In our case, that CSS selectors are `.second-hand` , `.min-hand` , `.hour-hand` respectively.

**What is `querySelector()`?** The `querySelector()` method returns the first element that matches a specified CSS class(s) in the document.

Learn more about [query selector](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector).

> Now it's time to implement a function for the proper functionality of the clock. Create a function `setDate` at the bottom your file.

Reviewer: ```suggestion
Now it's time to implement a function for the proper functionality of the clock. Create a function `setDate` at the bottom of your file.
```

  const minsDeg = ((mins / 60) * 360) + 90;
  minsHand.style.transform = `rotate(${minsDeg}deg)`;

  const hours = currentTime.getHours();
  const hoursDeg = ((hours / 12) * 360) + 90;
  hourHand.style.transform = `rotate(${hoursDeg}deg)`;

  if(seconds == 0){
    secondHand.style.transitionDuration = '0s';
    minsHand.style.transitionDuration = '0s';
    hourHand.style.transitionDuration = '0s';
  } else {
    secondHand.style.transitionDuration = '0.05s';
    minsHand.style.transitionDuration = '0.05s';
    hourHand.style.transitionDuration = '0.05s';
  }

  requestAnimationFrame(setDate);
}
setDate();
```

**And congratulations! You just built a clock using nearly 30 lines of JavaScript!**    

![Image](https://media.giphy.com/media/TdfyKrN7HGTIY/giphy.gif)

## Part 5: The End
If you haven't created an account on [repl.it](https://repl.it), make sure you do so to save this wonderful piece of creation!

Here's how you can create a repl.it account:

![Image](https://cloud-940o4yxjv.vercel.app/image.png)

![Image](https://cloud-18phelcmy.vercel.app/image.png)

If you still face difficulties in signing up watch [this](https://www.youtube.com/watch?v=Mtqp4CUepk0).

> Here are somethings which you can do:

Reviewer: ```suggestion
Here are some things which you can do:
```

1. Consider changing the colors and adding more features!  
2. Try to make the clock circular!  

> 3. Try to make it dark theme!

Reviewer: ```suggestion
3. Try to make it have a dark theme!
```
